core/services/workflows/syncer/v2/handler.go

```diff
@@ -0,0 +1,519 @@
+			h.lggr.Errorf("failed to emit status changed event: %+v", err)
+		}
+
+		h.lggr.Debugw("handled event", "workflowID", wfID, "workflowName", payload.WorkflowName, "workflowOwner", hex.EncodeToString(payload.WorkflowOwner), "type", event.Name)
+		return nil
+	case WorkflowDeleted:
+		payload, ok := event.Data.(WorkflowDeletedEvent)
+		if !ok {
+			return newHandlerTypeError(event.Data)
+		}
+
+		wfID := payload.WorkflowID.Hex()
+
+		cma := h.emitter.With(
+			platform.KeyWorkflowID, wfID,
+		)
+
+		if err := h.workflowDeletedEvent(ctx, payload); err != nil {
+			logCustMsg(ctx, cma, fmt.Sprintf("failed to handle workflow deleted event: %v", err), h.lggr)
+			return err
+		}
+
+		if err := events.EmitWorkflowStatusChangedEvent(ctx, h.emitter.Labels(), string(event.Name)); err != nil {
+			h.lggr.Errorf("failed to emit status changed event: %+v", err)
+		}
+
+		h.lggr.Debugw("handled event", "workflowID", wfID, "type", event.Name)
+		return nil
+	default:
+		return fmt.Errorf("event type unsupported: %v", event.Name)
+	}
+}
+
+// workflowRegisteredEvent handles the WorkflowRegisteredEvent event type.
+// This method must remain idempotent and must not error if retried multiple times.
+// workflowRegisteredEvent proceeds in two phases:
+// - phase 1 synchronizes the database state
+// - phase 2 synchronizes the state of the engine registry.
+func (h *eventHandler) workflowRegisteredEvent(
+	ctx context.Context,
+	payload WorkflowRegisteredEvent,
+) error {
+	owner := hex.EncodeToString(payload.WorkflowOwner)
+	status := toSpecStatus(payload.Status)
+
+	// First, let's synchronize the database state.
+	// We need to handle three cases:
+	// - new registration, without an existing DB record
+	// - existing registration that has been updated with new artifacts, and potentially also the status
+	// - existing registration that has been updated with a new status
+	spec, err := h.workflowArtifactsStore.GetWorkflowSpec(ctx, owner, payload.WorkflowName)
+	switch {
+	case err != nil:
+		newSpec, innerErr := h.createWorkflowSpec(ctx, payload)
+		if innerErr != nil {
+			return innerErr
+		}
+
+		spec = newSpec
+	case spec.WorkflowID != payload.WorkflowID.Hex():
+		newSpec, innerErr := h.createWorkflowSpec(ctx, payload)
+		if innerErr != nil {
+			return innerErr
+		}
+
+		spec = newSpec
+	case spec.Status != status:
+		spec.Status = status
+
+		if _, innerErr := h.workflowArtifactsStore.UpsertWorkflowSpec(ctx, spec); innerErr != nil {
+			return fmt.Errorf("failed to update workflow spec: %w", innerErr)
+		}
+	}
+
+	// Next, let's synchronize the engine.
+	// If the state isn't active, we shouldn't have an engine running.
+	// Let's try to clean one up if it exists
+	if spec.Status != job.WorkflowSpecStatusActive {
+		return h.tryEngineCleanup(payload.WorkflowID)
+	}
+
+	// We know we need an engine, let's make sure it's the right one.
+	// We do this by fetching and comparing whether it's running and that the workflow ID matches
```

Comment:
The comment hasn't been updated here I think -- we previously compared name + owner and checked if the id matched